src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java

```diff
@@ -0,0 +1,356 @@
+package net.snowflake.ingest.streaming.internal;
+
+import java.util.*;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import net.snowflake.ingest.streaming.InsertValidationResponse;
+import net.snowflake.ingest.streaming.OpenChannelRequest;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.Logging;
+import net.snowflake.ingest.utils.SFException;
+import net.snowflake.ingest.utils.Utils;
+import org.apache.arrow.util.VisibleForTesting;
+
+/**
+ * The abstract implementation of the buffer in the Streaming Ingest channel that holds the
+ * un-flushed rows, these rows will be converted to the underlying format implementation for faster
+ * processing
+ *
+ * @param <T> type of column data (Arrow {@link org.apache.arrow.vector.VectorSchemaRoot} or {@link
+ *     net.snowflake.ingest.streaming.internal.ParquetRowBuffer.ParquetChunkData})
+ */
+public abstract class AbstractRowBuffer<T> implements RowBuffer<T> {
+  private static final Logging logger = new Logging(AbstractRowBuffer.class);
+
+  // Snowflake table column logical type
+  enum ColumnLogicalType {
+    ANY,
+    BOOLEAN(1),
+    ROWINDEX,
+    NULL(15),
+    REAL(8),
+    FIXED(2),
+    TEXT(9),
+    CHAR,
+    BINARY(10),
+    DATE(7),
+    TIME(6),
+    TIMESTAMP_LTZ(3),
+    TIMESTAMP_NTZ(4),
+    TIMESTAMP_TZ(5),
+    INTERVAL,
+    RAW,
+    ARRAY(13, true),
+    OBJECT(12, true),
+    VARIANT(11, true),
+    ROW,
+    SEQUENCE,
+    FUNCTION,
+    USER_DEFINED_TYPE,
+    ;
+
+    private final int ordinal;
+    private final boolean object;
+
+    ColumnLogicalType() {
```

Comment:
Where is this used? Should we add a `ColumnLogicalType` of `INVALID` or `NOT_SET` corresponding to this value if we need it?

Reply:
I keep it mostly for readability to see what types cannot be actually packed into the chunk and read by server side.